File: op_tcg/frontend/components/decklist_builder.py
"""
Reusable decklist builder component.

The JavaScript logic lives in public/js/decklist_builder.js (loaded once at
page startup). This component only emits a tiny inline prefill script so HTMX
never has to re-execute the large IIFE inside swapped content.
"""
import json
from fasthtml import ft
from op_tcg.backend.db import get_decklist_watchlist, get_custom_decklists
from op_tcg.backend.models.cards import OPTcgCardCatagory


def create_decklist_builder(
    user_id: str,
    card_lookup: dict,
    *,
    custom_id: str = "",
    prefill_name: str = "",
    prefill_leader_id: str = "",
    prefill_leader_img: str = "",
    prefill_leader_name: str = "",
    prefill_decklist: dict | None = None,
) -> ft.Div:
    """
    Return the full builder HTML fragment (id='decklist-builder-wrapper').

    The caller is responsible for resolving prefill data from import params
    before passing it in. The JS logic is in decklist_builder.js.
    """
    if prefill_decklist is None:
        prefill_decklist = {}

    # ── Build card-keyed prefill dict ──────────────────────────────────────
    prefill_cards: dict = {}
    for card_id, count in prefill_decklist.items():
        c = card_lookup.get(card_id)
        prefill_cards[card_id] = {
            'count': int(count),
            'name': c.name if c else card_id,
            'img': c.image_url if c else '',
            'is_leader': (c.card_category == OPTcgCardCatagory.LEADER) if c else False,
        }

    # ── Leader select options (all leaders sorted newest-set-first then name) ──
    leaders_sorted = sorted(
        [c for c in card_lookup.values() if c.card_category == OPTcgCardCatagory.LEADER],
        key=lambda c: (c.meta_format or '', c.name),
        reverse=True,
    )
    leader_select_options = [ft.Option("— select a leader —", value="", disabled=True, selected=not prefill_leader_id)]
    for lc in leaders_sorted:
        leader_select_options.append(ft.Option(
            f"{lc.name} ({lc.id})",
            value=lc.id,
            selected=(lc.id == prefill_leader_id),
            data_leader_name=lc.name,
            data_leader_img=lc.image_url,
        ))

    # ── Import-source dropdown options ─────────────────────────────────────
    dl_watchlist = get_decklist_watchlist(user_id)
    custom_decklists_list = get_custom_decklists(user_id)

    import_options = [ft.Option("— import from watchlist —", value="", selected=True, disabled=True)]

    if dl_watchlist:
        import_options.append(ft.Option("── Tournament Decklists ──", value="", disabled=True))
        for item in dl_watchlist:
            lid = item.get('leader_id', '')
            lname = card_lookup[lid].name if lid in card_lookup else lid
            tid = item.get('tournament_id', '')
            pid = item.get('player_id', '')
            import_url = (
                f"/api/watchlist/custom-decklist/builder"
                f"?import_tournament_id={tid}&import_player_id={pid}"
                + (f"&custom_id={custom_id}" if custom_id else "")
            )
            import_options.append(ft.Option(
                f"{lname} — {tid[:30]}",
                value=f"t:{tid}:{pid}",
                data_import_url=import_url,
            ))

    if custom_decklists_list:
        import_options.append(ft.Option("── Custom Decklists ──", value="", disabled=True))
        for d in custom_decklists_list:
            if d.get('id') == custom_id:
                continue  # skip self when editing
            import_url = (
                f"/api/watchlist/custom-decklist/builder"
                f"?import_custom_id={d['id']}"
                + (f"&custom_id={custom_id}" if custom_id else "")
            )
            import_options.append(ft.Option(
                d.get('name', 'Unnamed'),
                value=f"c:{d['id']}",
                data_import_url=import_url,
            ))

    # ── Tiny prefill script (all logic is in decklist_builder.js) ──────────
    prefill_data = {
        'cards': prefill_cards,
        'leaderId': prefill_leader_id,
        'leaderName': prefill_leader_name,
        'leaderImg': prefill_leader_img,
        'customId': custom_id or None,
    }
    prefill_js = (
        f"window._cdbInit = {json.dumps(prefill_data)};"
        " if (typeof window._cdbSetup === 'function') window._cdbSetup();"
    )

    close_js = (
        "var w = document.getElementById('decklist-builder-wrapper');"
        " if (w) { w.innerHTML = ''; w.classList.add('hidden'); }"
    )

    return ft.Div(
        ft.Script(prefill_js),
        # ── Header ───────────────────────────────────────────────────────
        ft.Div(
            ft.H2(
                "Custom Decklist Builder" if not custom_id else "Edit Decklist",
                cls="text-base font-bold text-white",
            ),
            ft.Button(
                ft.I(cls="fas fa-times"),
                type="button",
                cls="text-gray-400 hover:text-white transition-colors ml-auto",
                onclick=close_js,
            ),
            cls="flex items-center gap-3 mb-4",
        ),
        # ── Name + Import row ─────────────────────────────────────────────
        ft.Div(
            ft.Div(
                ft.Label("Name", cls="text-xs text-gray-400 block mb-1"),
                ft.Input(
                    type="text", id="cdb-name", value=prefill_name,
                    placeholder="My awesome deck...",
                    cls="w-full bg-gray-700 text-white border border-gray-600 rounded"
                        " px-3 py-1.5 text-sm focus:outline-none focus:border-blue-500",
                ),
                cls="flex-1",
            ),
            ft.Div(
                ft.Label("Import from", cls="text-xs text-gray-400 block mb-1"),
                ft.Select(
                    *import_options,
                    id="cdb-import-select",
                    cls="w-full bg-gray-700 text-white border border-gray-600 rounded"
                        " px-2 py-1.5 text-sm focus:outline-none",
                    onchange="window._cdbImportChange(this)",
                ),
                cls="flex-1",
            ),
            cls="flex gap-3 mb-4 flex-col sm:flex-row",
        ),
        # ── Leader select + display ───────────────────────────────────────
        ft.Div(
            ft.Label("Leader", cls="text-xs text-gray-400 block mb-1"),
            ft.Select(
                *leader_select_options,
                id="cdb-leader-select",
                cls="w-full bg-gray-700 text-white border border-gray-600 rounded"
                    " px-2 py-1.5 text-sm focus:outline-none styled-select",
                onchange="window._cdbLeaderChange(this)",
            ),
            ft.Div(
                id="cdb-leader-display",
                cls="flex items-center gap-2 p-2 bg-gray-700/50 rounded border border-gray-600 min-h-[40px] mt-2",
            ),
            cls="mb-4",
        ),
        # ── Card search ───────────────────────────────────────────────────
        ft.Div(
            ft.Label("Search Cards", cls="text-xs text-gray-400 block mb-1"),
            ft.Input(
                type="text", name="search_term", id="cdb-search",
                placeholder="Search by name, meta format, type... (e.g. 'OP09 Luffy')",
                cls="w-full bg-gray-700 text-white border border-gray-600 rounded"
                    " px-3 py-1.5 text-sm focus:outline-none focus:border-blue-500 mb-2",
                hx_get="/api/decklist-builder/card-search",
                hx_trigger="keyup changed delay:400ms",
                hx_target="#cdb-search-results",
                hx_swap="innerHTML",
                hx_include="#cdb-search",
            ),
            ft.Div(
                ft.P("Type to search for cards.", cls="text-gray-500 text-sm text-center py-4"),
                id="cdb-search-results",
                cls="max-h-96 lg:max-h-[520px] overflow-y-auto",
            ),
            cls="mb-4",
        ),
        # ── Decklist panel ────────────────────────────────────────────────
        ft.Div(
            ft.Div(
                ft.Span("My Decklist", cls="text-xs text-gray-400 font-medium"),
                ft.Span("0 cards", id="cdb-deck-count", cls="text-xs text-gray-500 ml-2"),
                cls="flex items-center mb-2",
            ),
            ft.Div(
                ft.P("No cards yet.", cls="text-gray-500 text-sm text-center py-2"),
                id="cdb-decklist-panel",
                cls="max-h-48 overflow-y-auto",
            ),
            ft.Input(type="hidden", id="cdb-decklist-json", value="{}"),
            cls="mb-4 p-3 bg-gray-700/30 rounded border border-gray-600/50",
        ),
        # ── Footer ───────────────────────────────────────────────────────
        ft.Div(
            ft.Button(
                "Cancel",
                type="button",
                cls="px-4 py-2 text-sm text-gray-400 hover:text-white transition-colors",
                onclick=close_js,
            ),
            ft.Button(
                ft.I(cls="fas fa-save mr-2"),
                "Save Decklist",
                id="cdb-save-btn",
                type="button",
                cls="px-4 py-2 text-sm bg-blue-600 hover:bg-blue-500 text-white rounded"
                    " transition-colors font-medium inline-flex items-center",
                onclick="if (window._cdb) window._cdb.save();",
            ),
            cls="flex justify-end gap-3",
        ),
        id="decklist-builder-wrapper",
        cls="bg-gray-900 border border-gray-700 rounded-xl p-4 mb-6",
    )

File: public/js/decklist_builder.js
/**
 * Decklist Builder — static JS loaded once at page startup.
 * All _cdb logic lives here so HTMX never needs to re-execute it
 * inside swapped content (avoids normalizeScriptTags SyntaxErrors).
 *
 * The HTMX-swapped builder fragment only contains a tiny inline script:
 *   window._cdbInit = {...data...};
 *   if (typeof window._cdbSetup === 'function') window._cdbSetup();
 */
(function () {
  'use strict';

  /* ── Color-filter helper ────────────────────────────────────────────── */

  function setColorFilters(colors) {
    var el = document.getElementById('cdb-color-filters');
    if (!el) return;
    el.innerHTML = '';
    (colors || []).forEach(function (color) {
      var inp = document.createElement('input');
      inp.type = 'hidden';
      inp.name = 'card_colors';
      inp.value = color;
      el.appendChild(inp);
    });
  }

  /* ── Leader-select handler ──────────────────────────────────────────── */

  window._cdbLeaderChange = function (sel) {
    var opt = sel.options[sel.selectedIndex];
    if (!opt || !opt.value) return;
    var colors = [];
    try { colors = JSON.parse(opt.dataset.leaderColors || '[]'); } catch (e) {}
    if (window._cdb) {
      window._cdb.setLeader(opt.value, opt.dataset.leaderName || opt.value, opt.dataset.leaderImg || '', colors);
    }
  };

  /* ── Import-select handler ─────────────────────────────────────────── */

  window._cdbImportChange = function (sel) {
    var opt = sel.options[sel.selectedIndex];
    var url = opt ? opt.dataset.importUrl : null;
    if (!url) return;
    htmx.ajax('GET', url, { target: '#decklist-builder-wrapper', swap: 'innerHTML' });
  };

  /* ── Clipboard import handler ──────────────────────────────────────── */

  function _importCards(text) {
    if (!window._cdb) return;
    fetch('/api/decklist-builder/import-text', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ text: text })
    }).then(function (r) { return r.json(); })
    .then(function (cards) {
      if (!Array.isArray(cards) || cards.length === 0) {
        alert('No matching cards found. Check the format: 4xOP01-001');
        return;
      }
      var cdb = window._cdb;
      Object.keys(cdb.cards).forEach(function (k) {
        if (!cdb.cards[k].is_leader) delete cdb.cards[k];
      });
      cards.forEach(function (c) {
        if (c.is_leader) {
          cdb.setLeader(c.id, c.name || c.id, c.img || '', c.colors || []);
        } else {
          cdb.cards[c.id] = { count: c.count, name: c.name || c.id, img: c.img || '', is_leader: false };
        }
      });
      cdb.render();
    }).catch(function () {
      alert('Failed to import decklist. Please check the format and try again.');
    });
  }

  window._cdbPasteImport = function () {
    if (!window._cdb) return;
    navigator.clipboard.readText().then(function (text) {
      text = text.trim();
      if (!text) { alert('Clipboard is empty.'); return; }
      _importCards(text);
    }).catch(function () {
      alert('Could not read clipboard. Please allow clipboard access and try again.');
    });
  };

  /* ── Builder setup — called by the inline prefill script ───────────── */

  window._cdbSetup = function () {
    var init = window._cdbInit || {
      cards: {}, leaderId: '', leaderName: '', leaderImg: '', leaderColors: [], customId: null
    };
    window._cdbInit = null;

    window._cdb = {
      cards: JSON.parse(JSON.stringify(init.cards)),
      leaderId: init.leaderId,
      leaderName: init.leaderName,
      leaderImg: init.leaderImg,
      leaderColors: init.leaderColors || [],
      customId: init.customId,

      addFromBtn: function (btn) {
        var id = btn.dataset.cardId;
        var name = btn.dataset.cardName;
        var img = btn.dataset.cardImg;
        var isLeader = btn.dataset.isLeader === '1';
        if (isLeader) { this.setLeader(id, name, img, []); }
        else { this.addCard(id, name, img); }
      },

      addCard: function (id, name, img) {
        if (!(id in this.cards)) {
          this.cards[id] = { count: 0, name: name, img: img, is_leader: false };
        }
        if (this.cards[id].count < 4) this.cards[id].count++;
        this.render();
      },

      removeCard: function (id) {
        if (!(id in this.cards)) return;
        this.cards[id].count--;
        if (this.cards[id].count <= 0) delete this.cards[id];
        this.render();
      },

      setLeader: function (id, name, img, colors) {
        var cdb = this;
        Object.keys(this.cards).forEach(function (k) {
          if (cdb.cards[k].is_leader) delete cdb.cards[k];
        });
        this.leaderId = id;
        this.leaderName = name;
        this.leaderImg = img;
        this.leaderColors = colors || [];
        this.cards[id] = { count: 1, name: name, img: img, is_leader: true };
        setColorFilters(this.leaderColors);
        this.renderLeader();
        this.render();
      },

      renderLeader: function () {
        var el = document.getElementById('cdb-leader-display');
        if (!el) return;
        if (this.leaderId) {
          el.innerHTML =
            '<img src="' + this.leaderImg + '" class="w-10 h-auto rounded" alt="">' +
            '<span class="text-white text-sm font-medium">' + this.leaderName + '</span>' +
            '<span class="text-xs text-gray-400 ml-1">(' + this.leaderId + ')</span>';
        } else {
          el.innerHTML =
            '<span class="text-gray-400 text-sm">No leader selected — use the dropdown above</span>';
        }
      },

      render: function () {
        var panel = document.getElementById('cdb-decklist-panel');
        var countEl = document.getElementById('cdb-deck-count');
        var hidden = document.getElementById('cdb-decklist-json');
        var cards = this.cards;
        var totalCards = Object.values(cards).reduce(function (s, c) { return s + c.count; }, 0);
        if (countEl) countEl.textContent = totalCards + ' cards';

        var decklistOut = {};
        Object.entries(cards).forEach(function (e) { decklistOut[e[0]] = e[1].count; });
        if (hidden) hidden.value = JSON.stringify(decklistOut);
        if (!panel) return;

        if (Object.keys(cards).length === 0) {
          panel.innerHTML = '<p class="text-gray-500 text-sm text-center py-2">No cards yet.</p>';
          return;
        }

        panel.innerHTML = '';
        Object.entries(cards).forEach(function (entry) {
          var id = entry[0];
          var data = entry[1];
          var row = document.createElement('div');
          row.className = 'flex items-center justify-between py-1 border-b border-gray-700/30 last:border-0';

          /* Escape single quotes so onclick strings stay valid */
          var safeId   = id.replace(/\\/g, '\\\\').replace(/'/g, "\\'");
          var safeName = (data.name || id).replace(/\\/g, '\\\\').replace(/'/g, "\\'");
          var safeImg  = (data.img || '').replace(/\\/g, '\\\\').replace(/'/g, "\\'");

          row.innerHTML =
            '<div class="flex items-center flex-1 min-w-0 mr-2">' +
              '<img src="' + (data.img || '') + '" class="w-7 h-auto rounded mr-2 flex-shrink-0" alt="">' +
              '<span class="text-white text-xs truncate">' + (data.name || id) + '</span>' +
              (data.is_leader ? '<span class="ml-1 text-xs text-yellow-400">(Leader)</span>' : '') +
            '</div>' +
            '<div class="flex items-center flex-shrink-0">' +
              '<button onclick="window._cdb.removeCard(\'' + safeId + '\')" ' +
                'class="w-6 h-6 bg-gray-600 hover:bg-gray-500 text-white rounded-l text-sm leading-none flex items-center justify-center">' +
                '\u2212</button>' +
              '<span class="w-7 text-center text-white text-xs bg-gray-700 h-6 flex items-center justify-center">' +
                data.count + '</span>' +
              '<button onclick="window._cdb.addCard(\'' + safeId + '\',\'' + safeName + '\',\'' + safeImg + '\')" ' +
                'class="w-6 h-6 bg-gray-600 hover:bg-gray-500 text-white rounded-r text-sm leading-none flex items-center justify-center">' +
                '+</button>' +
            '</div>';
          panel.appendChild(row);
        });
      },

      save: function () {
        var cdb = this;
        var nameEl = document.getElementById('cdb-name');
        var name = nameEl ? nameEl.value.trim() : '';
        if (!name) { alert('Please enter a decklist name.'); return; }
        if (!cdb.leaderId) { alert('Please select a leader card.'); return; }
        var jsonEl = document.getElementById('cdb-decklist-json');
        var decklistJson = jsonEl ? jsonEl.value : '{}';
        var btn = document.getElementById('cdb-save-btn');
        if (btn) {
          btn.disabled = true;
          btn.querySelector('i').className = 'fas fa-spinner fa-spin mr-2';
        }
        fetch('/api/watchlist/custom-decklist/save', {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({
            name: name,
            leader_id: cdb.leaderId,
            decklist: JSON.parse(decklistJson),
            custom_id: cdb.customId || null
          })
        }).then(function (r) {
          if (r.ok) {
            window.location.href = '?section=decklists';
          } else {
            alert('Failed to save. Please try again.');
            if (btn) {
              btn.disabled = false;
              btn.querySelector('i').className = 'fas fa-save mr-2';
            }
          }
        }).catch(function () {
          alert('Error saving decklist.');
          if (btn) {
            btn.disabled = false;
            btn.querySelector('i').className = 'fas fa-save mr-2';
          }
        });
      }
    };

    /* Apply prefill color filters and render */
    setColorFilters(window._cdb.leaderColors);
    window._cdb.renderLeader();
    window._cdb.render();
  };
})();
